interfaces/prompting/requestprompts/requestprompts.go

```diff
@@ -0,0 +1,373 @@
+	userEntry, ok := pdb.perUser[user]
+	if !ok || len(userEntry.ByID) == 0 {
+		return []*Prompt{}
+	}
+	prompts := make([]*Prompt, 0, len(userEntry.ByID))
+	for _, prompt := range userEntry.ByID {
+		prompts = append(prompts, prompt)
+	}
+	return prompts
+}
+
+// PromptWithID returns the prompt with the given ID for the given user.
+func (pdb *PromptDB) PromptWithID(user uint32, id string) (*Prompt, error) {
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+	userEntry, ok := pdb.perUser[user]
+	if !ok || len(userEntry.ByID) == 0 {
+		return nil, fmt.Errorf("cannot find prompt for UID %d with the given ID: %s", user, id)
+	}
+	prompt, ok := userEntry.ByID[id]
+	if !ok {
+		return nil, fmt.Errorf("cannot find prompt for UID %d with the given ID: %s", user, id)
+	}
+	return prompt, nil
+}
+
+// Reply resolves the prompt with the given ID using the given outcome by
+// sending a reply to all associated listener requests, then removing the
+// prompt from the prompt DB.
+//
+// Records a notice for the prompt, and returns the prompt's former contents.
+func (pdb *PromptDB) Reply(user uint32, id string, outcome prompting.OutcomeType) (*Prompt, error) {
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+	userEntry, ok := pdb.perUser[user]
+	if !ok || len(userEntry.ByID) == 0 {
+		return nil, fmt.Errorf("cannot find prompt for UID %d with the given ID: %s", user, id)
+	}
+	prompt, ok := userEntry.ByID[id]
+	if !ok {
+		return nil, fmt.Errorf("cannot find prompt for UID %d with the given ID: %s", user, id)
+	}
+	allow, err := outcome.IsAllow()
+	if err != nil {
+		return nil, err
+	}
+	for _, listenerReq := range prompt.listenerReqs {
+		if err := sendReply(listenerReq, allow); err != nil {
+			return nil, err
+		}
+	}
+	delete(userEntry.ByID, id)
+	pdb.notifyPrompt(user, id)
+	return prompt, nil
+}
+
+var sendReply = func(listenerReq *listener.Request, reply interface{}) error {
+	return listenerReq.Reply(reply)
+}
+
+// HandleNewRule checks if any existing prompts are satisfied by the given rule
+// contents and, if so, sends back a decision to their listener requests.
+//
+// A prompt is satisfied by the given rule contents if the user, snap,
+// interface, and path of the prompt match those of the rule, and if either the
+// outcome is "allow" and all of the prompt's permissions are matched by those
+// of the rule contents, or if the outcome is "deny" and any of the permissions
+// match.
+//
+// Records a notice for any prompt which was satisfied, or which had some of
+// its permissions satisfied by the rule contents. In the future, only the
+// remaining unsatisfied permissions of a partially-satisfied prompt must be
+// satisfied for the prompt as a whole to be satisfied.
+//
+// Returns the IDs of any prompts which were fully satisfied by the given rule
+// contents.
+func (pdb *PromptDB) HandleNewRule(user uint32, snap string, iface string, constraints *prompting.Constraints, outcome prompting.OutcomeType) ([]string, error) {
```

Comment:
AddOrMerge and this could share a struct for some of the arguments here

Reply:
User, snap, and interface are shared across almost all prompting-related things, I'll create a `prompting.Metadata` struct with these fields.

The remaining ones left out of the Metadata or Constraints structs would then be outcome, lifespan, duration. I think that's fine, we can revisit later if need be.